File: training/training/report/new_joinees_details/new_joinees_details.js
// Copyright (c) 2016, Ramya and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["New Joinees Details"] = {
	"filters": [
        {
			"fieldname": "date_of_joining_from",
			"label": __("Date of Joining From"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1
        },
        {
			"fieldname": "date_of_joining_to",
			"label": __("Date of Joining To"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1
        },
    ],
    "formatter": function (row, cell, value, columnDef, dataContext, default_formatter) {
        if (columnDef.id == "Induction Test Assy") {
            value = dataContext['Induction Test Assy']
            columnDef.df.link_onclick =
				"frappe.query_reports['New Joinees Details'].open_att_adjust(" + JSON.stringify(dataContext) + ")"; 
        }
        if (columnDef.id == "Induction Test Machine") {
            value = dataContext['Induction Test Machine']
            columnDef.df.link_onclick =
				"frappe.query_reports['New Joinees Details'].open_att_adjust1(" + JSON.stringify(dataContext) + ")"; 
        }
        if (columnDef.id == "Selection Test") {
            value = dataContext['Selection Test']
            columnDef.df.link_onclick =
				"frappe.query_reports['New Joinees Details'].open_att_adjust3(" + JSON.stringify(dataContext) + ")"; 
        }
        if (columnDef.id == "Line") {
            value = dataContext['Line']
            columnDef.df.link_onclick =
				"frappe.query_reports['New Joinees Details'].open_att_adjust2(" + JSON.stringify(dataContext) + ")"; 
        }
        if (columnDef.id == "Shift") {
            value = dataContext['Shift']
            columnDef.df.link_onclick =
				"frappe.query_reports['New Joinees Details'].open_att_adjust4(" + JSON.stringify(dataContext) + ")"; 
        }
        if (columnDef.id == "Color Blindness Test") {
            value = dataContext["Color Blindness Test"]
            columnDef.df.link_onclick =
				"frappe.query_reports['New Joinees Details'].open_att_adjust5(" + JSON.stringify(dataContext) + ")"; 
        }
        if (columnDef.id == "Practical Knowledge Verification") {
            value = dataContext["Practical Knowledge Verification"]
            columnDef.df.link_onclick =
				"frappe.query_reports['New Joinees Details'].open_att_adjust6(" + JSON.stringify(dataContext) + ")"; 
        }
        value = default_formatter(row, cell, value, columnDef, dataContext);
    return value;
    },
    "open_att_adjust": function (data) {
        if(data['Induction Test Assy'] =="Pending"){
            frappe.set_route("Form","Induction Training Assembly Area","New Induction Training Assembly Area",{"employee_code": data["Employee"]});
        }
        if(data['Induction test Assy'] == "Completed"){
            frappe.set_route("Form","Induction Training Assembly Area",data["Employee"]);
        }
    },

    "open_att_adjust1": function (data) {
        if(data['Induction Test Machine'] == "Pending"){
            frappe.set_route("Form","Induction Training Machine Area Crimping","New Induction Training Machine Area Crimping",{"employee_code": data["Employee"]});
        } 
        if(data['Induction Test Machine'] == "Completed"){
            frappe.set_route("Form","Induction Training Machine Area Crimping",data["Employee"]);
        }
    },
    "open_att_adjust3": function (data) {
        if(data['Selection Test'] == "Pending"){
            frappe.set_route("Form","Selection Test","New Selection Test",{"employee_code": data["Employee"]});
        }
        if(data['Selection Test'] == "Completed"){
            frappe.set_route("Form","Selection Test",data["Employee"]);
        }
    },
    "open_att_adjust2": function (data) {
        // console.log(data)
        if(data['Line'] == "Pending"){
            frappe.set_route("Form","Line and Shift Assignment Tool","New Line and Shift Assignment Tool",{"employee": data["Employee"]});
        }
        if(data['Line'] == "new"){
            frappe.set_route("Form","Line and Shift Assignment Tool","New Line and Shift Assignment Tool",{"employee": data["Employee"]});
        }
        else if(data['Line']){
            frappe.set_route("Form","Line and Shift Assignment Tool","New Line and Shift Assignment Tool",{"employee": data["Employee"]});
        }
    },
    "open_att_adjust4": function (data) {
        if(data['Shift'] == "Pending"){
            frappe.set_route("Form","Line and Shift Assignment Tool","New Line and Shift Assignment Tool",{"employee": data["Employee"]});
        }
        else if(data['Shift']){
            frappe.set_route("Form","Line and Shift Assignment Tool","New Line and Shift Assignment Tool",{"employee": data["Employee"]});
        }
    },
    "open_att_adjust5": function (data) {
        if(data['Color Blindness Test'] == "Pending"){
            frappe.set_route("Form","Color Blindness Test","New Color Blindness Test",{"employee_code": data["Employee"]});
        }
        if(data['Color Blindness Test'] == "Completed"){
            frappe.set_route("Form","Color Blindness Test",data["Employee"]);
        }
    },
    "open_att_adjust6": function (data) {
        if(data['Practical Knowledge Verification'] == "Pending"){
            frappe.set_route("Form","New Joinees Practical Knowledge Verification","New New Joinees Practical Knowledge Verification",{"employee_code": data["Employee"]});
        }
        if(data['Practical Knowledge Verification'] == "Completed"){
            frappe.set_route("Form","New Joinees Practical Knowledge Verification",data["Employee"]);
        }
    }
}




l

File: training/training/report/new_joinees_details/new_joinees_details.py
# Copyright (c) 2013, Minda Sai Pvt LTd and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
import math
from calendar import monthrange
from datetime import datetime,timedelta,date
from frappe.utils import add_days
from dateutil.rrule import * 

def execute(filters=None):
    if not filters:
        filters = {}
        
    columns = get_columns()

    data = []
    row = []
    employee = employee_details(filters)
    for e in employee:
        row = [e.biometric_id,e.employee_name,e.date_of_joining]
        att = att_details(e.biometric_id)
        ita = ita_details(e.biometric_id)
        itm = itm_details(e.biometric_id)
        st = st_details(e.biometric_id)
        cb = cb_details(e.biometric_id)
        pk = pk_details(e.biometric_id)

        if e.line == None:
            row += ["Pending"]
        else:
            row += [e.line]
        if e.shift not in ["A","B","C","G"]:
            row += ["Pending"]
        else:
            row += [e.shift]
        if att == 1:
            row += ["Active"]
        else:
            row += ["Continuous Absent"]
        if e.department not in ["HR", "Accounts","Finance & Accounts","Purchase"]:
            if cb == 1:
                row += ["Completed"]
            else:
                row += ["Pending"]
        else:
            row += ["NA"]
        if st == 1:            
            row += ["Completed"]
        else:
            row += ["Pending"]
        if e.department not in ["HR", "Accounts","Finance & Accounts","Purchase"]:
            if pk == 1:
                row += ["Completed"]
            else:
                row += ["Pending"]
        else:
            row += ["NA"]
        if e.department not in ["HR", "Accounts","Finance & Accounts","Purchase"]:
            if ita == 1:       
                row += ["Completed","NA"]
            elif itm == 1:
                row += ["NA","Completed"]
            else:
                row += ["Pending","Pending"]
        else:
            row += ["NA"]
        data.append(row)
    return columns, data




def get_columns():
    columns = [
        _("Employee") + ":Link/Employee:50",
        _("Employee Name") + ":Data:100",
        _("Date of Joining") + ":Date:100",
        _("Line") + ":Link/Line:100",
        _("Shift") + ":Link/Shift:100",
        _("Status") + ":Select:100",
        _("Color Blindness Test") + ":Link/Color Blindness Test:100",
        _("Selection Test") + ":Link/Selection Test:100",
        _("Practical Knowledge Verification") + ":Link/New Joinees Practical Knowledge Verification:100",
        _("Induction Test Assy") + ":Link/Induction Training Assembly Area:100",
        _("Induction Test Machine") + ":Link/Induction Training Machine Area Crimping:100",
        
    ]
    return columns


# def get_conditions(filters):
#     conditions = ""
#     # if filters.get("employee"):conditions += "AND att.employee = '%s'" % filters["employee"]
#     if filters.get("from_date"): conditions += "and c.date_of_skill_evaluatation >= %(from_date)s"
#     if filters.get("to_date"): conditions += " and c.date_of_skill_evaluatation <= %(to_date)s"    
#     return conditions, filters


def employee_details(filters):
    employee = frappe.db.sql(
        """select biometric_id,employee_name,shift,status,department,designation,date_of_joining,line from `tabEmployee` where date_of_joining between %s and %s and status = 'Active' """,(filters.get("date_of_joining_from"),filters.get("date_of_joining_to")),as_dict = 1)
    return employee

def ita_details(emp_id):
    if emp_id:
        ita = frappe.db.sql(
            """select * from `tabInduction Training Assembly Area` where employee_code=%s """,(emp_id),as_dict = 1)
        if ita:
            return True
        else:
            return False

def itm_details(emp_id):
    if emp_id:
        ita = frappe.db.sql(
            """select * from `tabInduction Training Machine Area Crimping` where employee_code=%s """,(emp_id),as_dict = 1)
        if ita:
            return True
        else:
            return False

def att_details(emp_id):
    today = date.today()
    yesterday = add_days(today,-1)
    day_before_yesterday = add_days(today,-2)
    if emp_id:
        att = frappe.db.sql(
            """select attendance_date from `tabAttendance` where employee = %s and attendance_date in (%s,%s,%s)  """,(emp_id,today,yesterday,day_before_yesterday),as_dict = 1)
        if len(att) > 0:
            return True
        else:
            return False
def st_details(emp_id):
    if emp_id:
        st = frappe.db.sql(
            """select * from `tabSelection Test` where employee_code=%s """,(emp_id),as_dict = 1)
        if st:
            return True
        else:
            return False

def cb_details(emp_id):
    if emp_id:
        cb = frappe.db.sql(
            """select * from `tabColor Blindness Test` where employee_code=%s """,(emp_id),as_dict = 1)
        if cb:
            return True
        else:
            return False

def pk_details(emp_id):
    if emp_id:
        pk = frappe.db.sql(
            """select * from `tabNew Joinees Practical Knowledge Verification` where employee_code=%s """,(emp_id),as_dict = 1)
        if pk:
            return True
        else:
            return False